### DISCUSSION TITLE
document.head - refresh

### DISCUSSION THREAD

**Discussion by BurningTreeC** [TiddlyWiki/TiddlyWiki5#8128] [Ideas]
While trying to get each single Stylesheet into a single `<style>` tag I found that I can modify the $:/core/templates/tiddlywiki5.html tiddler to transclude a "HEAD template" tiddler.

I then found out that I can create another "HEAD template" tiddler (it needs a slightly different form) and make a `transclude` widget in `render.js` from it and render it in the document.head section.
I get all stylesheets as single `<style>` tags and can refresh the whole `document.head` section when something changes.
What I cannot achieve is getting the comments into the head section:

```
<!--~~ Raw markup for the top of the head section ~~-->
```

Is there an option for the `makeTranscludeWidget` method not to strip those comments?
Or is there another way to get them into the head section?

I also didn't really find the place in the core code where the $:/core/templates/tiddlywiki5.html tiddler gets rendered.
Currently I replace the `document.head` with the new rendered head which is very clumsy.

**Comment by Jermolene** [comment]
Hi @BurningTreeC I think perhaps the comments would need to be in backticks so that TiddlyWiki doesn't parse them as TW comments.

**Reply by BurningTreeC** [reply]
Hi @Jermolene, no the backticks make them end up wrapped within `<code>` tags.
I'm now using `\rules except commentinline commentblock` for that purpose and get the comments in the head section.
But they're not green like the original ones.

**Comment by BurningTreeC** [comment]
How is $:/core/templates/tiddlywiki5.html parsed? I'd like to resemble that parsing

**Reply by Jermolene** [reply]
It's generally used via `$:/core/save/all` which transcludes it. In the browser you can see it being used in `$:/core/modules/saver-handler.js`, where  it is rendered with the output taken as text/plain.

**Reply by BurningTreeC** [reply]
Ah thanks @Jermolene, I see!
One question: can we have something like rules to skip the `<code>` tags generated when we use backticks?
Like `\skip codeinline`?
That would be very useful here

**Reply by BurningTreeC** [reply]
By `skip` I mean that the `<code>` tags are not generated but the text is rendered as text, hot htmlified

**Reply by Jermolene** [reply]
Hi @BurningTreeC that sounds a wee bit of a kludge. Can you share what you've got so far?

**Reply by BurningTreeC** [reply]
Hi @Jermolene 

I've got this in render.js to get the `$:/core/ui/HeadTemplate` into the `document.head`

```
	var parser = $tw.wiki.parseTiddler("$:/core/ui/HeadTemplate",{parseAsInline: true}),
		widgetNode = $tw.wiki.makeWidget(parser,{document: document});
	widgetNode.render(document.head,null);

	$tw.wiki.addEventListener("change",$tw.perf.report("styleRefresh",function(changes) {
		widgetNode.refresh(changes);
	}));
```

Then I have this in the `$:/core/ui/HeadTemplate`

```
title: $:/core/ui/HeadTemplate

\import [subfilter{$:/core/config/GlobalImportFilter}]
\define saveTiddlerFilter()
[is[tiddler]] -[prefix[$:/state/popup/]] -[prefix[$:/temp/]] -[prefix[$:/HistoryList]] -[status[pending]plugin-type[import]] -[[$:/boot/boot.css]] -[type[application/javascript]library[yes]] -[[$:/boot/boot.js]] -[[$:/boot/bootprefix.js]] +[sort[title]] $(publishFilter)$
\end
{{$:/core/ui/HeadSubTemplate}}
```

I transclude the `$:/core/ui/HeadSubTemplate`, so:

```
title: $:/core/ui/HeadSubTemplate

<$set name="saveTiddlerAndShadowsFilter" filter="[subfilter<saveTiddlerFilter>] [subfilter<saveTiddlerFilter>plugintiddlers[]]">
<$transclude tiddler="$:/core/ui/RootStylesheet" mode="inline"/>
<meta http-equiv="Content-Type" content="text/html;charset=utf-8" />
<!--~~ Raw markup for the top of the head section ~~-->
{{{ [enlist<saveTiddlerAndShadowsFilter>tag[$:/tags/RawMarkupWikified/TopHead]] ||$:/core/templates/raw-static-tiddler}}}
<meta http-equiv="X-UA-Compatible" content="IE=Edge"/>
<meta name="application-name" content="TiddlyWiki" />
<meta name="generator" content="TiddlyWiki" />
<$wikify name="core-templates-version" text={{$:/core/templates/version}}>
<meta name="tiddlywiki-version" content=<<core-templates-version>> />
</$wikify>
<meta name="viewport" content="width=device-width, initial-scale=1.0" />
<meta name="apple-mobile-web-app-capable" content="yes" />
<meta name="apple-mobile-web-app-status-bar-style" content="black-translucent" />
<meta name="mobile-web-app-capable" content="yes"/>
<meta name="format-detection" content="telephone=no" />
<meta name="copyright" content={{$:/core/copyright.txt}}/>
<$wikify name="datauri" text="""<$macrocall $name="makedatauri" text={{$:/favicon.ico}} type={{$:/favicon.ico!!type}}/>""">
<link id="faviconLink" rel="shortcut icon" href=<<datauri>>>
</$wikify>
<title><$view tiddler="$:/core/wiki/title" format="plainwikified"/></title>
<!--~~ This is a Tiddlywiki file. The points of interest in the file are marked with this pattern ~~-->

<!--~~ Raw markup ~~-->
{{{ [enlist<saveTiddlerAndShadowsFilter>tag[$:/core/wiki/rawmarkup]] ||$:/core/templates/plain-text-tiddler}}}
{{{ [enlist<saveTiddlerAndShadowsFilter>tag[$:/tags/RawMarkup]] ||$:/core/templates/plain-text-tiddler}}}
{{{ [enlist<saveTiddlerAndShadowsFilter>tag[$:/tags/RawMarkupWikified]] ||$:/core/templates/raw-static-tiddler}}}
</$set>
```

And finally, the `$:/core/ui/HeadSubTemplate` transcludes the `$:/core/ui/RootStylesheet`

```
title: $:/core/ui/RootStylesheet
code-body: yes

\whitespace trim
<$set name="currentTiddler" value={{$:/language}}>
<$set name="languageTitle" value={{!!name}}>
<$list filter="[all[shadows+tiddlers]tag[$:/tags/Stylesheet]!has[draft.of]]">
<style type="text/css">
<$view format="plainwikified"/>
</style>
</$list>
</$set>
</$set>
```

**Reply by BurningTreeC** [reply]
If we remove the `favicon.js` startup module this works like a charm.
But having to use wikify widgets is a big downside.
And The comments are not there

**Reply by BurningTreeC** [reply]
I could render into the fakeDocument and use `.innerHTML` and `.textContent` but that would mean that I'd have to replace the complete `innerHTML` of the head each time something changes. It's faster though doing so

**Reply by Jermolene** [reply]
Hi @BurningTreeC it is unfortunate that we have to change so much about the way that unrelated parts of the document `<head>` are handled. It is necessary because we need to be generating an actual `<style>` DOM node, not just outputting the source HTML.

Having said that, this implementation is in many ways cleaner because it avoids the backticks that make some TW templates very hard to follow.

Is this new template intended to also be used for rendering on the server?

**Reply by BurningTreeC** [reply]
Hi @Jermolene, first of all I'd like to find a way to actually use the backticks.

To your question about rendering on the server, do you mean by the `--render` command or `--rendertiddlers` ?
I've seen that there it nees the backticks since it renders into the fakeDom.
I'm trying if I find a way to render into the document but keeping the backticks.

**Reply by BurningTreeC** [reply]
... Or the other way around

**Reply by pmario** [reply]
>But having to use wikify widgets is a big downside.
And The comments are not there

I'm not really sure, what's done here. But I'm pretty sure that we do not need `wikify` for core templates.

**Reply by Jermolene** [reply]
Hi @BurningTreeC we have two types of templates in TiddlyWiki:

* Ordinary templates, where any HTML elements are written natively
* Plain text templates, where any HTML elements are written as plain text (typically using backticks)

The second type of template always produces a string of HTML text. In TW, that HTML is always saved to a file or delivered to the server. TW doesn't currently render those templates to the DOM, but it could only do so by setting innerHTML (or similar), to get the browser the parse the HTML.

I added these plain text templates to TW as a bit of a kludge, but in practice these templates can often be quite convenient to work with. They have something of the flavour of templating languages like PHP. From a certain perspective, your work on this topic has been laying the ground for removing the need for plain text templates.

So, it is now clear that this isn't a quick, drop-in optimisation. This is a fairly major piece of refactoring, touching on a mechanism that needs to work consistently in the browser and on Node.js. It would also not be backwards compatible, so we'd have to keep the old templates around so that old code continues to work.

**Reply by BurningTreeC** [reply]
Hi @pmario - instead of the `styleRefresh` we make a `headRefresh` where the whole `<head>` is a widget node and gets simply refreshed

**Reply by BurningTreeC** [reply]
Hi @Jermolene 

Yes I've found out about these two different template types and their up- and downsides.
No this is not a quick drop-in optimisation.
I've also come to the conclusion, that the existing templates need to stay.

How do you think about such a `headRefresh`?
It seems to me that it's quite heavy but it works

**Reply by Jermolene** [reply]
> How do you think about such a `headRefresh`? It seems to me that it's quite heavy but it works

I'm concerned that there is too much duplication: it would mean that we have two different templates for generating the `<head>` section in the browser, which I think will be error prone and brittle.

**Reply by BurningTreeC** [reply]
Is there a filter we can evaluate for detecting browser or nodejs?

**Reply by BurningTreeC** [reply]
Yes there is, but it's not the way to go

**Comment by pmario** [comment]
>How is `$:/core/templates/tiddlywiki5.html` parsed? I'd like to resemble that parsing

- `$:/core/templates/tiddlywiki5.html` is part of [$:/core/save/all](https://github.com/Jermolene/TiddlyWiki5/blob/e43cd2d989bd564fa983ad0a28369345e20855d7/core/templates/save-all.tid#L7)
- `$:/core/save/all` is the default root-tiddler as shown for the for a [TW server help](https://tiddlywiki.com/#%24%3A%2Flanguage%2FHelp%2Flisten)
- `$:/core/save/all` is used as a parameter for the `tm-download-file` message in the [Download full wiki](https://tiddlywiki.com/#%24%3A%2Fcore%2Fui%2FDownloadFullWiki) button
- `tm-download-file` is handled by: [saver-handler.js](https://github.com/Jermolene/TiddlyWiki5/blob/e43cd2d989bd564fa983ad0a28369345e20855d7/core/modules/saver-handler.js#L103) and calls
- [self.saveWiki](https://github.com/Jermolene/TiddlyWiki5/blob/e43cd2d989bd564fa983ad0a28369345e20855d7/core/modules/saver-handler.js#L151) which does a 
  - saveWiki `options` are important
- [text = this.wiki.renderTiddler(downloadType,template,options),](https://github.com/Jermolene/TiddlyWiki5/blob/e43cd2d989bd564fa983ad0a28369345e20855d7/core/modules/saver-handler.js#L163)
- I think that's basically it. 

hope that helps

**Reply by BurningTreeC** [reply]
Thank you @pmario !

**Comment by BurningTreeC** [comment]
I have one question @Jermolene @pmario 

Do the `<style>` tags need to be on top of the `<head>` section or can they also be on the bottom?
